tests/pyanaconda_tests/kickstart_dispatcher.py

```diff
@@ -0,0 +1,514 @@
+text
+%addon
+anaconda
+%end
+""".lstrip()
+
+
+class TrackedKickstartElementsTest(unittest.TestCase):
+
+    def tracked_kickstart_elements_test(self):
+
+        expected_ks_content = """
+%pre
+echo PRE
+%end
+network --device=ens3 --activate
+network --device=ens4 --activate
+%addon pony --fly=True
+%end
+firewall --enabled
+%addon scorched --planet=Earth
+nuke
+%end
+%post --nochroot --interpreter /usr/bin/bash
+echo POST1
+%end
+""".lstrip()
+
+        filename = "ks.test.simple.cfg"
+
+        element1 = KickstartElement(("%pre",), "%pre\necho PRE\n%end\n", 1, filename)
+        element2 = KickstartElement(("network",), "network --device=ens3 --activate\n", 4, filename)
+        element3 = KickstartElement(("network",), "network --device=ens4 --activate\n", 5, filename)
+        element4 = KickstartElement(("%addon", "pony"), "%addon pony --fly=True\n%end\n", 6, filename)
+        element5 = KickstartElement(("firewall",), "firewall --enabled\n", 8, filename)
+        element6 = KickstartElement(("%addon", "scorched"), "%addon scorched --planet=Earth\nnuke\n%end\n", 9, filename)
+        element7 = KickstartElement(("%post",), "%post --nochroot --interpreter /usr/bin/bash\necho POST1\n%end\n", 12, filename)
+        expected_elements = [element1, element2, element3, element4, element5, element6, element7]
+
+        elements = TrackedKickstartElements()
+        elements.append_section(element1.args[0], element1.content, element1.lineno, filename)
+        elements.append_command(element2.args[0], element2.content, element2.lineno, filename)
+        elements.append_command(element3.args[0], element3.content, element3.lineno, filename)
+        elements.append_addon(element4.args[1], element4.content, element4.lineno, filename)
+        elements.append_command(element5.args[0], element5.content, element5.lineno, filename)
+        elements.append_addon(element6.args[1], element6.content, element6.lineno, filename)
+        elements.append_section(element7.args[0], element7.content, element7.lineno, filename)
+
+        ### Filtering elements
+
+        all_elements = elements.all_elements
+        self.assertEqual(all_elements, expected_elements)
+
+        # elements of specified type and filter are returned
+        network_commands = elements.get_elements(commands=["network"])
+        self.assertEqual(network_commands, [element2, element3])
+        pony_addon = elements.get_elements(addons=["pony"])
+        self.assertEqual(pony_addon, [element4])
+        self.assertEqual(elements.get_elements(), [])
+
+        # to get all elements of specified type use empty list as filter
+        all_addons_and_sections = elements.get_elements(sections=[], addons=[])
+        self.assertEqual(all_addons_and_sections, [element1, element4, element6, element7])
+
+        ### Tracking elements
+
+        processed_elements = elements.get_elements(commands=["network"], sections=["%pre"], addons=["pony"], process=True)
```

Comment:
It doesn't look very consistent, that a section has to start with `%` and an addon is missing `%addon`. Maybe modules could specify the kickstart elements with their "full" names. Then it could look like: 

`elements.get_elements(["network", "%pre", "%addon pony"], process=True)`

Then we don't have to care about the types of the elements we have and just keep their "full" names in `KickstartElement`. It will simplify the method `get_elements` as well.

Or we can just drop `%` from the section names.

Reply:
You are right.
I think I'll drop the % and try to work just with names and (typed) elements.

The
`elements.get_elements(["network", "%pre", "%addon pony"], process=True)`
seems dangerously low-level for me thinking about it now (like client needs to construct "%addon pony" if she wants a "pony" addon). We need to decompose / compose the "%addon pony" anyway when working with kickstart as the parser gives us only already parsed (shlex) args of section header (not the line itself) in the method overriden by the splitter. So I think we shouldn't work with kickstart line beginnings containing implicit element type info but with some elementary abstractions (element types:) - which would be also safer for future if some more complicated element is added to pykickstart.

Reply:
Ok.